File: src/main/java/miragefairy2019/mod/modules/fairyweapon/items/ItemFairyWandSummoning.java
package miragefairy2019.mod.modules.fairyweapon.items;

import java.util.List;

import miragefairy2019.mod.api.ApiFairy.EnumAbilityType;
import miragefairy2019.mod.api.Components;
import miragefairy2019.mod.modules.fairycrystal.ItemFairyCrystal;
import miragefairy2019.mod.modules.fairyweapon.ItemFairyCraftingToolBase;
import net.minecraft.client.util.ITooltipFlag;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.item.EntityItem;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.EnumAction;
import net.minecraft.item.ItemStack;
import net.minecraft.stats.StatList;
import net.minecraft.util.ActionResult;
import net.minecraft.util.EnumActionResult;
import net.minecraft.util.EnumHand;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.RayTraceResult;
import net.minecraft.util.math.RayTraceResult.Type;
import net.minecraft.util.text.TextFormatting;
import net.minecraft.world.World;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;

public class ItemFairyWandSummoning extends ItemFairyCraftingToolBase
{

	public ItemFairyWandSummoning()
	{
		composite = composite
			.add(Components.MIRAGIUM, 1)
			.add(Components.MOONSTONE, 1)
			.add(Components.fairyAbilityType(EnumAbilityType.crystal));
		setMaxDamage(64 - 1);
	}

	@Override
	@SideOnly(Side.CLIENT)
	public void addInformation(ItemStack itemStack, World world, List<String> tooltip, ITooltipFlag flag)
	{

		// ポエム
		tooltip.add("餌付けのステッキ");

		super.addInformation(itemStack, world, tooltip, flag);

	}

	@Override
	@SideOnly(Side.CLIENT)
	protected void addInformationFunctions(ItemStack itemStack, World world, List<String> tooltip, ITooltipFlag flag)
	{

		tooltip.add(TextFormatting.RED + "Hold right mouse button to use fairy crystals quickly");

		super.addInformationFunctions(itemStack, world, tooltip, flag);

	}

	//

	@Override
	public EnumAction getItemUseAction(ItemStack stack)
	{
		return EnumAction.BOW;
	}

	// 永続
	@Override
	public int getMaxItemUseDuration(ItemStack stack)
	{
		return 72000;
	}

	@Override
	public ActionResult<ItemStack> onItemRightClick(World world, EntityPlayer player, EnumHand hand)
	{
		player.setActiveHand(hand);
		return new ActionResult<ItemStack>(EnumActionResult.SUCCESS, player.getHeldItem(hand));
	}

	@Override
	public void onUsingTick(ItemStack stack, EntityLivingBase entityLivingBase, int count)
	{
		if (entityLivingBase.world.isRemote) return;

		if (entityLivingBase instanceof EntityPlayer) {
			EntityPlayer player = (EntityPlayer) entityLivingBase;

			// 使用Tickじゃないなら抜ける
			if (!isUsingTick(count)) return;

			// 妖晶を得る
			ItemStack nItemStackFairyCrystal = findItem(player, itemStack -> itemStack.getItem() instanceof ItemFairyCrystal).orElse(null);
			if (nItemStackFairyCrystal == null) return;

			// プレイヤー視点判定
			RayTraceResult rayTraceResult = rayTrace(player.world, player, false);
			if (rayTraceResult == null) return; // ブロックに当たらなかった場合は無視
			if (rayTraceResult.typeOfHit != Type.BLOCK) return; // ブロックに当たらなかった場合は無視

			// ガチャを引く
			ItemStack nItemStackDrop = ItemFairyCrystal.drop(
				player,
				player.world,
				rayTraceResult.getBlockPos(),
				player.getHeldItem(EnumHand.MAIN_HAND).getItem() == this ? EnumHand.MAIN_HAND : EnumHand.OFF_HAND,
				rayTraceResult.sideHit,
				(float) rayTraceResult.hitVec.x,
				(float) rayTraceResult.hitVec.y,
				(float) rayTraceResult.hitVec.z).orElse(null);
			if (nItemStackDrop == null) return; // ガチャが引けなかった場合は無視
			if (nItemStackDrop.isEmpty()) return; // ガチャが引けなかった場合は無視

			// 成立

			// ガチャアイテムを消費
			nItemStackFairyCrystal.shrink(1);
			player.addStat(StatList.getObjectUseStats(nItemStackFairyCrystal.getItem()));

			// 妖精をドロップ
			BlockPos blockPos = rayTraceResult.getBlockPos().offset(rayTraceResult.sideHit);
			EntityItem entityItem = new EntityItem(player.world, blockPos.getX() + 0.5, blockPos.getY() + 0.5, blockPos.getZ() + 0.5, nItemStackDrop.copy());
			entityItem.setNoPickupDelay();
			player.world.spawnEntity(entityItem);

		}
	}

	private boolean isUsingTick(int count)
	{
		count = 72000 - count;
		if (count >= 60) return true;
		if (count >= 20) return count % 2 == 0;
		if (count >= 5) return count % 5 == 0;
		if (count == 1) return true;
		return false;
	}

}

File: src/main/java/miragefairy2019/mod/modules/fairyweapon/item/ItemFairyWandSummoning.kt
package miragefairy2019.mod.modules.fairyweapon.item

import miragefairy2019.mod.modules.fairycrystal.ItemFairyCrystal
import miragefairy2019.mod.modules.fairyweapon.item.FairyWeaponUtils.findItemOptional
import miragefairy2019.mod3.skill.api.ApiSkill
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.item.EntityItem
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.EnumAction
import net.minecraft.item.ItemStack
import net.minecraft.stats.StatList
import net.minecraft.util.ActionResult
import net.minecraft.util.EnumActionResult
import net.minecraft.util.EnumHand
import net.minecraft.util.math.RayTraceResult
import net.minecraft.world.World

class ItemFairyWandSummoning(val maxTryCountPerTick: Int) : ItemFairyWand() {

    override val magicFeatureInformation get() = "右クリック長押しでフェアリークリスタルを高速消費" // TODO translate Hold right mouse button to use fairy crystals quickly

    //

    // TODO フェアリーステッキクラフト無効
    //override fun onItemUse(player: EntityPlayer, worldIn: World, pos: BlockPos, hand: EnumHand, facing: EnumFacing, hitX: Float, hitY: Float, hitZ: Float) = EnumActionResult.PASS
    //override fun onUpdate(itemStack: ItemStack, world: World, entity: Entity, itemSlot: Int, isSelected: Boolean) = Unit

    override fun getItemUseAction(stack: ItemStack) = EnumAction.BOW
    override fun getMaxItemUseDuration(stack: ItemStack) = 72000 // 永続

    override fun onItemRightClick(world: World, player: EntityPlayer, hand: EnumHand): ActionResult<ItemStack> {
        player.activeHand = hand
        return ActionResult(EnumActionResult.SUCCESS, player.getHeldItem(hand))
    }

    override fun onUsingTick(stack: ItemStack, entityLivingBase: EntityLivingBase, count: Int) {
        if (entityLivingBase.world.isRemote) return

        if (entityLivingBase is EntityPlayer) {
            repeat(getTryCount(count)) a@{
                if (!tryUseCrystal(entityLivingBase)) return@a
            }
        }
    }

    private fun tryUseCrystal(player: EntityPlayer): Boolean {

        // 妖晶を得る
        val itemStackFairyCrystal = findItemOptional(player) { itemStack -> itemStack!!.item is ItemFairyCrystal }.orElse(null) ?: return false // クリスタルを持ってない場合は無視
        val variantFairyCrystal = (itemStackFairyCrystal.item as ItemFairyCrystal).getVariant(itemStackFairyCrystal) ?: return false // 異常なクリスタルを持っている場合は無視

        // プレイヤー視点判定
        val rayTraceResult = rayTrace(player.world, player, false) ?: return false // ブロックに当たらなかった場合は無視
        if (rayTraceResult.typeOfHit != RayTraceResult.Type.BLOCK) return false // ブロックに当たらなかった場合は無視

        // ガチャを引く
        val itemStackDrop = variantFairyCrystal.dropper.drop(
            player,
            player.world,
            rayTraceResult.blockPos,
            if (player.getHeldItem(EnumHand.MAIN_HAND).item === this) EnumHand.MAIN_HAND else EnumHand.OFF_HAND,
            rayTraceResult.sideHit,
            rayTraceResult.hitVec.x.toFloat(),
            rayTraceResult.hitVec.y.toFloat(),
            rayTraceResult.hitVec.z.toFloat(),
            variantFairyCrystal.dropRank,
            variantFairyCrystal.getRareBoost(ApiSkill.skillManager.getServerSkillContainer(player))
        ) ?: return false // ガチャが引けなかった場合は無視
        if (itemStackDrop.isEmpty) return false // ガチャが引けなかった場合は無視

        // 成立

        // ガチャアイテムを消費
        if (!player.isCreative) itemStackFairyCrystal.shrink(1)
        player.addStat(StatList.getObjectUseStats(itemStackFairyCrystal.item))

        // 妖精をドロップ
        val blockPos = rayTraceResult.blockPos.offset(rayTraceResult.sideHit)
        val entityItem = EntityItem(player.world, blockPos.x + 0.5, blockPos.y + 0.5, blockPos.z + 0.5, itemStackDrop.copy())
        entityItem.setNoPickupDelay()
        player.world.spawnEntity(entityItem)

        return true
    }

    private fun getTryCount(count: Int): Int {
        val t = 72000 - count
        return when {
            t >= 200 -> 5
            t >= 100 -> 2
            t >= 60 -> 1
            t >= 20 -> if (t % 2 == 0) 1 else 0
            t >= 5 -> if (t % 5 == 0) 1 else 0
            t == 1 -> 1
            else -> 0
        }.coerceAtMost(maxTryCountPerTick)
    }
}
